consensus/safety-rules/src/safety_rules.rs

```diff
@@ -98,6 +99,37 @@ impl<T: Payload> SafetyRules<T> {
         }
     }
 
+    /// This verifies the public key and private key of a validator against validator set
+    /// during epoch changes. We might want to add more detailed error code for identifying
+    /// the failure of key verification
+    fn verify_key(&mut self, epoch_state: &EpochState) -> Result<(), Error> {
+        // the validator set
+        let validator_verifier = &epoch_state.verifier;
+        let validator_signer = self
+            .validator_signer
+            .as_ref()
+            .ok_or(Error::NotInitialized)?;
+
+        // validator account not in the validator set
+        let public_key = validator_signer.public_key();
+        let author = validator_signer.author();
+        let validator_key = validator_verifier
+            .get_public_key(&author)
+            .ok_or(Error::InvalidAuthor)?;
```

Comment:
This is confusing, the error implies the author of a proposal doesn't exist within the validator verifier, perhaps we forgot to include this check in sign proposal :O

Reply:
I also find this tricky and what I do here is I'm trying to catch the possiblity that the signer is not in the validator set. And my understanding is sign_proposal() and epoch changing is two different thing; the signer verification being passed in sign_proposal (i.e. with verify_qc) cannot transfer during epoch changes. Therefore I tried to propagate the error ahd place some TODOs to see how that could be handled (I tried to panic!() but it does not work).

Reply:
the naming is confusing, we have invalid author as one of the verification error. this one means the node itself is not a validator anymore

Reply:
also i prefer to just panic in such a case instead of returning error